File: native-views/react-native-native-list/ios/NativeListScrollPositionTracker.swift
import Foundation

/// A per-view hysteresis observer. A nil result means no bridge event is needed.
struct NativeListScrollPositionTracker {
  private var start: Double?
  private var end: Double?
  private var previous: Bool?

  mutating func configure(json: String) {
    start = nil
    end = nil
    previous = nil
    guard let data = json.data(using: .utf8),
          let values = try? JSONSerialization.jsonObject(with: data) as? [String: Double],
          let start = values["start"], let end = values["end"],
          start.isFinite, end.isFinite, start >= 0, end > start else { return }
    self.start = start
    self.end = end
  }

  mutating func resetDelivery() { previous = nil }

  mutating func update(offset: Double) -> Bool? {
    guard let start, let end, offset.isFinite else { return nil }
    let next = previous == true ? offset > start : offset >= end
    guard previous != next else { return nil }
    previous = next
    return next
  }
}

File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListScrollPositionTracker.kt
package com.margelo.nitro.nativelist

/** A per-view hysteresis observer. Null means no bridge event is needed. */
internal class NativeListScrollPositionTracker {
  private var start: Double? = null
  private var end: Double? = null
  private var previous: Boolean? = null

  fun configure(start: Double?, end: Double?) {
    this.start = null
    this.end = null
    previous = null
    if (start == null || end == null || !start.isFinite() || !end.isFinite() || start < 0 || end <= start) return
    this.start = start
    this.end = end
  }

  fun resetDelivery() { previous = null }

  fun update(offset: Double): Boolean? {
    val start = start ?: return null
    val end = end ?: return null
    if (!offset.isFinite()) return null
    val next = if (previous == true) offset > start else offset >= end
    if (previous == next) return null
    previous = next
    return next
  }
}
